### PR TITLE
MGMT-21201: Enable dual-stack clusters with image-based installations / upgrades

### PR DESCRIPTION
**Pull request by danmanor**
### 🎯 Overview
This PR adds comprehensive dual-stack networking support to OpenShift image-based installations, enabling clusters to use both IPv4 and IPv6 networks simultaneously. This addresses the growing need for dual-stack deployments in modern containerized environments.

#### Dual-Stack Network Configuration Support
- **Multiple Machine Networks**: Extended support from single `MachineNetwork` to multiple `MachineNetworks` array
- **IPv4 + IPv6 Support**: Full support for configuring both IPv4 and IPv6 machine networks
- **Validation Logic**: Added comprehensive validation to ensure dual-stack configurations are valid (max 2 networks, one IPv4 + one IPv6)

#### Smart DHCP Kernel Arguments
- **Automatic DHCP Detection**: Intelligently determines when to add DHCP kernel arguments based on network configuration
- **Conditional Logic**: Only adds kernel args for dual-stack scenarios (single-stack remains unchanged)
- **User Override Protection**: Respects user-provided kernel arguments and avoids duplication
- **Network Config Parsing**: Analyzes NetworkManager state configuration to determine static vs. DHCP for each IP version

#### Enhanced Data Structures
- **Updated SeedReconfiguration**: Added `MachineNetworks []string` field while maintaining backward compatibility with deprecated `MachineNetwork`
- **Install Config Validation**: Extended validation to support dual-stack machine network configurations

#### Comprehensive Testing
- **Install Config Tests**: 189+ lines of new test cases covering dual-stack scenarios
- **Ignition Tests**: 354+ lines of new tests including DHCP kernel argument logic
- **Test Infrastructure**: Improved test setup with proper resource management
- **Edge Cases**: Coverage for mixed static/DHCP, user-provided args, and validation scenarios

#### DHCP Kernel Arguments Logic
```go
// Only adds kernel args for dual-stack scenarios
if len(machineNetworks) == 2 && hasIPv4 && hasIPv6 {
    // Analyzes networkConfig to determine what needs DHCP
    // Returns: "ip=dhcp", "ip=dhcp6", "ip=dhcp,dhcp6", or ""
}
```

#### Machine Networks Migration
```go
// Before: Single network only
cc.Config.MachineNetwork = networks[0].CIDR.String()

// After: Support for multiple networks
funk.ForEach(networks, func(net *types.MachineNetworkEntry) {
    cc.Config.MachineNetworks = append(cc.Config.MachineNetworks, net.CIDR.String())
})
```

### ✅ Validation & Testing
- ✅ All existing tests pass
- ✅ New dual-stack test scenarios added
- ✅ Backward compatibility maintained
- ✅ Edge cases covered (single-stack, invalid configs, user overrides)

### 🔄 Backward Compatibility
- Maintains full backward compatibility with existing single-stack configurations
- Deprecated `MachineNetwork` field still supported but `MachineNetworks` takes precedence
- No breaking changes to existing APIs or configurations

### CODE REVIEW

File: pkg/types/imagebased/seedreconfiguration.go [hunk 1]
```diff
@@ -61,11 +61,28 @@ type SeedReconfiguration struct {
 	// be empty or match the first entry in the list.
 	// Default is 10.0.0.0/16 for all platforms other than Power VS.
 	// For Power VS, the default is 192.168.0.0/24.
+	// Use MachineNetworks instead.
```
Comment (eranco74) [inline pkg/types/imagebased/seedreconfiguration.go]: // Deprecated: Use MachineNetworks instead.

File: pkg/types/imagebased/seedreconfiguration.go [hunk 2]
```diff
@@ -61,11 +61,28 @@ type SeedReconfiguration struct {
...
+	// For single stack ocp clusters, this will be a single subnet.
+	// For dual-stack ocp clusters, this will be a list of two subnets.
+	// This will be used to create the node network and choose ip addresses for the node.
+	// Equivalent to install-config.yaml's machineNetworks.
+	// Use this field instead of MachineNetwork.
+	// If both MachineNetwork and MachineNetworks are specified, MachineNetworks takes precedence.
+	MachineNetworks []string `json:"machine_networks,omitempty"`
+
 	// NodeIP is the desired IP address of the node.
+	// Use NodeIPs instead.
```
Comment (eranco74) [inline pkg/types/imagebased/seedreconfiguration.go]: // Deprecated: Use NodeIPs instead.

File: pkg/asset/imagebased/image/ignition.go [hunk 3]
```diff
@@ -39,6 +42,7 @@ func (i *Ignition) Dependencies() []asset.Asset {
 		&ImageBasedInstallationConfig{},
 		&RegistriesConf{},
 		&PostDeployment{},
+		&configimage.InstallConfig{},
```
Comment (mresvanis) [inline pkg/asset/imagebased/image/ignition.go]: Does this mean that in order to create the IBI installation ISO we now require the user to prepare an `install-config.yaml` ? 

If yes, then I think  we need to put whatever fields we need from the `install-config.yaml` into the `imagebased-installation-config.yaml`.  

I understand that it's not super clear by looking into the respective [design doc](https://github.com/openshift/enhancements/blob/master/enhancements/installer/image-based-installer.md) (or just the codebase), but having an `install-config.yaml` as input for both the create installation ISO and create config ISO subcommands leads to a confusing user experience and we [decided](https://github.com/openshift/enhancements/pull/1617#discussion_r1654313325) that it fits only to the create config ISO subcommand.

**Comment by mresvanis** [comment]
Thanks for the contribution @danmanor! Do you think we can add the respective integration tests for this change?

The integration tests for the `imagebased` subcommand can be found [here](https://github.com/openshift/installer/tree/3474a3e90f7840818b616994539c5d6b79365f62/cmd/openshift-install/testdata/imagebased). These tests are based on [github.com/rogpeppe/go-internal/testscript](github.com/rogpeppe/go-internal/testscript) and can be executed locally via `./hack/go-integration-test.sh`.

File: pkg/asset/imagebased/image/ignition.go (hunk 3 above)
Comment (danmanor) [reply pkg/asset/imagebased/image/ignition.go]: yes I understand. we have 2 options then:
1. Not adding `dhcp,dhcp6` kernel args automatically, ask the user to specify it in the docs for dual-stack if needed.
2. Add IP stack information to image-based-config. 
@mresvanis WDYT ? Honestly I am not sure it is needed

File: pkg/asset/imagebased/image/ignition.go [hunk 4]
```diff
@@ -181,3 +193,127 @@ func defaultEnabledServices() []string {
...
+		// Also check if someone provided ip= directly (though unusual)
+		if strings.HasPrefix(arg, "ip=") {
+			return true
+		}
+	}
+	return false
+}
+
+// isIPVersionStatic checks if a specific IP version is configured for static networking.
+func isIPVersionStatic(networkConfigStr, ipVersion string) bool {
```
Comment (mresvanis) [inline pkg/asset/imagebased/image/ignition.go]: What do you think about unmarshalling the `networkConfigStr` to a struct and using that one to do all the checks (it's YAML anyway), instead of manipulating lines of strings?

The following is Just an example (it surely needs more work), but it would look like this:

```go
var config map[string]interface{}
if err := yaml.Unmarshal([]byte(networkConfigStr), &config); err != nil {
    return fmt.Errorf("failed to parse network config as YAML: %w", err)
}

// Extract interfaces configuration
interfaces, ok := config["interfaces"]
if !ok {
    return nil
}

interfacesList, ok := interfaces.([]interface{})
if !ok {
    return nil
}

// Analyze each interface for IPv4 and IPv6 configuration
ipv4NeedsDHCP := true
ipv6NeedsDHCP := true

for _, iface := range interfacesList {
    ifaceMap, ok := iface.(map[interface{}]interface{})
    if !ok {
	continue
    }

    // Check IPv4 configuration
    if ipv4Config, exists := ifaceMap["ipv4"]; exists {
	if ipv4Map, ok := ipv4Config.(map[interface{}]interface{}); ok {
		if hasStaticAddress(ipv4Map) && !isDHCPEnabled(ipv4Map) {
			ipv4NeedsDHCP = false
		}
	}
    }

    // Check IPv6 configuration
    if ipv6Config, exists := ifaceMap["ipv6"]; exists {
	if ipv6Map, ok := ipv6Config.(map[interface{}]interface{}); ok {
		if hasStaticAddress(ipv6Map) && !isDHCPEnabled(ipv6Map) {
			ipv6NeedsDHCP = false
		}
	}
    }
}
```

File: pkg/asset/imagebased/image/ignition.go (hunk 3 above)
Comment (mresvanis) [reply pkg/asset/imagebased/image/ignition.go]: Just to clarify what I wrote above, we do have the `networkConfig` field in the [IBI installation config](https://github.com/openshift/installer/blob/3474a3e90f7840818b616994539c5d6b79365f62/pkg/types/imagebased/imagebased_config_types.go#L102) so I think we should fetch it from there, no need for the `install-config.yaml`.

Please let me know if you think I'm missing anything.
